### PR TITLE
Application & Subscription Sharing Implementions

### PR DESCRIPTION
**Pull request by shaniR**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 1]
```diff
@@ -2576,7 +2580,13 @@ public static NativeArray jsFunction_getAllSubscriptions(Context cx,
 
             Subscriber subscriber = new Subscriber(username);
             APIConsumer apiConsumer = getAPIConsumer(thisObj);
-            Application[] applications = apiConsumer.getApplications(new Subscriber(username));
+            Application[] applications;
+            if("".equals(groupId) || groupId == null){
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Use isEmpty() with refactoring this as if(groupId==null || groupId.isEmpty())

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 2]
```diff
@@ -2593,9 +2603,13 @@ public static NativeArray jsFunction_getAllSubscriptions(Context cx,
 	                    appName.equals(application.getName())) {
 
 		                //get subscribed APIs set for application
-		                Set<SubscribedAPI> subscribedAPIs =
-				                apiConsumer.getSubscribedAPIs(subscriber, application.getName());
-
+	                	Set<SubscribedAPI> subscribedAPIs;
+                        if (groupId == null || "".equals(groupId)) {
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Use .isEmpty()

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 3]
```diff
@@ -2897,7 +2914,17 @@ public static NativeArray jsFunction_getApplications(Context cx,
         if (args != null && isStringArray(args)) {
             String username = args[0].toString();
             APIConsumer apiConsumer = getAPIConsumer(thisObj);
-            Application[] applications = apiConsumer.getApplications(new Subscriber(username));
+            Application[] applications;
+            String groupId="";
+            if(args.length >1){
+            	 groupId = args[1].toString();
+            }
+            if ("".equals(groupId) || groupId == null) {
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Use isEmpty() with refactoring this as if(groupId==null || groupId.isEmpty())

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 4]
```diff
@@ -2920,27 +2947,39 @@ public static String jsFunction_addApplication(Context cx,
...
             APIConsumer apiConsumer = getAPIConsumer(thisObj);
             Subscriber subscriber = new Subscriber(username);
 
-            Application[] apps = apiConsumer.getApplications(subscriber);
-            for (Application app : apps) {
-                if (app.getName().equals(name)) {
-                    handleException("A duplicate application already exists by the name - " + name);
-                }
+            Application[] apps = null;
+            if (groupId == null || "".equals(groupId)) {
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Use isEmpty() with refactoring this as if(groupId==null || groupId.isEmpty())

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 5]
```diff
@@ -4139,4 +4194,27 @@ public static NativeObject jsFunction_getDocument(Context cx, Scriptable thisObj
...
+     * 
+     */
+    public static String jsFunction_getGroupIds(Context cx, Scriptable thisObj, Object[] args, Function funObj) {
+        String response = (String) args[0];
+        APIConsumer consumer = getAPIConsumer(thisObj);
+        String groupId = null;
+        try {
+            groupId = consumer.getGroupIds(response);
+        } catch (APIManagementException e) {
+            log.error("Error occurred while getting group id");
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: put a comment,mentioning the reason,why swallowing the exception..

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 6]
```diff
@@ -997,10 +997,21 @@ public void addSubscriber(Subscriber subscriber) throws APIManagementException {
                 subscriberId = Integer.valueOf(rs.getString(1)).intValue();
             }
             subscriber.setId(subscriberId);
+            String groupId = subscriber.getGroupId();
+
+            // Check for duplicate default applications if group id is available.
+            if (groupId != "" && groupId != null) {
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Use isEmpty() with refactoring this as if(groupId!=null && !groupId.isEmpty())

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 7]
```diff
@@ -2565,6 +2565,10 @@ public static NativeArray jsFunction_getAllSubscriptions(Context cx,
         try {
             String username = args[0].toString();
             String appName = args[1].toString();
+            String groupId = null;
+            if(args.length > 2){
+            	groupId = args[2].toString();
```
Comment (lalaji) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Better to handle NPE..

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 8]
```diff
@@ -2971,12 +3010,18 @@ public static boolean jsFunction_removeApplication(Context cx,
...
-        if (args != null && isStringArray(args)) {
+        if (args != null && args.length > 2 && isStringArray(args)) {
             String name = (String) args[0];
             String username = (String) args[1];
+            String groupId = (String) args[2];
             Subscriber subscriber = new Subscriber(username);
             APIConsumer apiConsumer = getAPIConsumer(thisObj);
-            Application[] apps = apiConsumer.getApplications(subscriber);
+            Application[] apps;
+            if(groupId != null || groupId != ""){
```
Comment (abimarank) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Can you please use !isEmpty() method?

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 9]
```diff
@@ -3036,17 +3081,27 @@ public static boolean jsFunction_updateApplication(Context cx,
...
+            if(args.length > 6 && args[6] != null){
+            	groupId = (String) args[6];
+            }
+           
             Subscriber subscriber = new Subscriber(username);
             APIConsumer apiConsumer = getAPIConsumer(thisObj);
-            Application[] apps = apiConsumer.getApplications(subscriber);
-            if (apps == null || apps.length == 0) {
+            Application[] apps;
+            if(groupId == null || groupId == "" ){
```
Comment (abimarank) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Here too, Can you please use isEmpty() method?

**Pull request merged**